include/up-cpp/uri/serializer/LongUriSerializer.h

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright (c) 2023 General Motors GTO LLC
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * 
+ * SPDX-FileType: SOURCE
+ * SPDX-FileCopyrightText: 2023 General Motors GTO LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef LONG_URI_SERIALIZER_H_
+#define LONG_URI_SERIALIZER_H_
+
+#include <up-cpp/uri/builder/BuildUResource.h>
+#include <up-cpp/uri/builder/BuildEntity.h>
+#include <up-cpp/uri/builder/BuildUAuthority.h>
+#include <up-cpp/uri/builder/BuildUUri.h>
+#include <up-cpp/uri/tools/Utils.h>
+#include <up-core-api/uri.pb.h>
+#include <string_view>
+
+namespace uprotocol::uri {
+
+/**
+ * UUri Serializer that serializes a UUri to a string (long format) per
+ * https://github.com/eclipse-uprotocol/uprotocol-spec/blob/main/basics/uri.adoc
+ */
+class LongUriSerializer {
+public:
+    /**
+     * Support for serializing UUri objects into their String format.
+     *
+     * @param uUri UUri object to be serialized to the String format.
+     *
+     * @return Returns the String format of the supplied UUri that can be used
+     * as a sink or a source in a uProtocol publish communication.
+     */
+    static auto serialize(const v1::UUri& uri) -> std::string;
```

Comment:
Interesting question :thinking: 

Over in `up-rust`, we completely dismantled the idea of having separate serializers and deserializers.

We instead leaned on the Rust trait `TryFrom` and `impl`ed that on the `UUri` struct to allow going:
* `UUri` -> `String`, `String` -> `UUri` (i.e. long form) and
* `UUri` -> `Vec<u8>`, `Vec<u8>` -> `UUri` (i.e. micro form)

which, thinking about it now, is a bit of a fumble, since both Long Form and Short Form are `String`s, so we'll need to revisit this

Reply:
Yeah, the string nature of the two formats does sort of break the idea of automatic conversions (at least, without a little type ✨magic✨). I am not sure that C++ has a direct parallel to the `TryFrom`, but that might be my inexperience with rust. Maybe [`boost::lexical_cast`](https://www.boost.org/doc/libs/1_85_0/doc/html/boost_lexical_cast.html) comes close?

I think really my point is that making this a `class` isn't providing any benefit since that class is really only used as a glorified `namespace`. We could just have namespaced functions and it would be easier to read.

Reply:
I've settled on tiny classes in namespaces for serialize / deserialize. They are designed such that they could be passed as template parameters to functions. An example is in UMessageBuilder, where payloads can be passed to build along with a serializer class.

This is [utilized in the updated example repo](https://github.com/gregmedd/up-zenoh-example-cpp/blob/feature/new-api-demo/rpc/src/main_rpc_server.cpp#L57) ([serializer](https://github.com/gregmedd/up-zenoh-example-cpp/blob/feature/new-api-demo/rpc/include/common.h#L55)).